src/dependency.cr

```diff
@@ -1,60 +1,83 @@
 require "./ext/yaml"
 
 module Shards
-  class Dependency < Hash(String, String)
+  class Dependency
     property name : String
+    setter version : String?
+    property! resolver : String?
+    property! url : String?
+    property tag : String?
+    property branch : String?
+    property commit : String?
+
+    getter unmapped = Hash(String, YAML::Any).new
 
     def self.new(pull : YAML::PullParser) : self
-      Dependency.new(pull.read_scalar).tap do |dependency|
-        pull.each_in_mapping do
-          dependency[pull.read_scalar] = pull.read_scalar
+      dependency = Dependency.new(pull.read_scalar)
+
+      pull.each_in_mapping do
+        case key = pull.read_scalar
+        when "version"
+          dependency.version = pull.read_scalar
+        when "url"
+          dependency.url = pull.read_scalar
+        when "tag"
+          dependency.tag = pull.read_scalar
+        when "branch"
+          dependency.branch = pull.read_scalar
+        when "commit"
+          dependency.commit = pull.read_scalar
+        when "path", "git", "github", "gitlab", "bitbucket"
```

Comment:
Just mapping the names would be problematic when multiple mappings are specified in a single entry. Per current spec that's valid. When you only have the names you later don't know about the order of definition for picking the last one as the final one (which is IMO a sane option to treat this, and the current implementation in this PR).

Besides that, the data model is much easier and more efficient when you can simply retrieve the name of the resolver and its url (or "resolution value"), than having to iterate through the possible resolvers in order to fine which one's specified. I'd be happy to change `resolver` to `resolver_name`. Both `owner/repo` and local paths are URLs, so technically that's not incorrect. I'm feeling quite strongly that this is a superior solution over multiple properties.

Reply:
- Its forbidden to have multiple resolvers for a single dependency declaration. SPEC can evolve to make that explicit. It makes no sense to have multiple resolvers.
- There is no order in dependency keys, either.
- `owner/repo` isn't an URL (barely a path) and path isn't an URL either but an absolute path (it would be `file://`).

Reply:
I'll change it to error if multiple resolvers are mapped.

Technically, relative paths are a URI reference. But naming aside, do you agree on normalized properties for resolver name and reference?

Reply:
@ysbaddaden ?